File: src/core/engine-detection.js
import { isVerboseEnabled } from './logger.js';

/**
 * Detect which browser automation engine is being used
 * @param {Object} pageOrContext - Page or context object from Playwright or Puppeteer
 * @returns {string} - 'playwright' or 'puppeteer'
 */
export function detectEngine(pageOrContext) {
  const hasEval = !!pageOrContext.$eval;
  const hasEvalAll = !!pageOrContext.$$eval;
  const locatorType = typeof pageOrContext.locator;
  const contextType = typeof pageOrContext.context;
  const hasContext = contextType === 'function' || contextType === 'object';

  // Debug logging
  if (isVerboseEnabled()) {
    console.log('🔍 [ENGINE DETECTION]', {
      hasEval,
      hasEvalAll,
      locatorType,
      contextType,
      hasContext,
    });
  }

  // Check for Playwright-specific methods first
  // Playwright has locator as a function and context() method
  // Both engines have $eval and $$eval, so we check for unique Playwright features first
  if (locatorType === 'function' && hasContext) {
    if (isVerboseEnabled()) {
      console.log('🔍 [ENGINE DETECTION] Detected: playwright');
    }
    return 'playwright';
  }
  // Check for Puppeteer-specific methods
  // Puppeteer has $eval, $$eval but no context() method
  if (hasEval && hasEvalAll && !hasContext) {
    if (isVerboseEnabled()) {
      console.log('🔍 [ENGINE DETECTION] Detected: puppeteer');
    }
    return 'puppeteer';
  }
  if (isVerboseEnabled()) {
    console.log('🔍 [ENGINE DETECTION] Could not detect engine!');
  }
  throw new Error(
    'Unknown browser automation engine. Expected Playwright or Puppeteer page object.'
  );
}

File: python/src/browser_commander/core/engine_detection.py
"""Engine detection for browser automation frameworks."""

from typing import Any, Literal

from browser_commander.core.logger import is_verbose_enabled

EngineType = Literal["playwright", "selenium"]


def detect_engine(page_or_driver: Any) -> EngineType:
    """Detect which browser automation engine is being used.

    from __future__ import annotations

        Args:
            page_or_driver: Page or driver object from Playwright or Selenium

        Returns:
            Engine type: 'playwright' or 'selenium'

        Raises:
            ValueError: If the engine cannot be detected
    """
    # Check for Playwright-specific attributes
    has_locator = hasattr(page_or_driver, "locator") and callable(
        getattr(page_or_driver, "locator", None)
    )
    has_context = hasattr(page_or_driver, "context")
    has_goto = hasattr(page_or_driver, "goto") and callable(
        getattr(page_or_driver, "goto", None)
    )

    # Check for Selenium-specific attributes
    has_find_element = hasattr(page_or_driver, "find_element") and callable(
        getattr(page_or_driver, "find_element", None)
    )
    has_get = hasattr(page_or_driver, "get") and callable(
        getattr(page_or_driver, "get", None)
    )
    has_current_url = hasattr(page_or_driver, "current_url")

    if is_verbose_enabled():
        print(
            f"[ENGINE DETECTION] has_locator={has_locator}, "
            f"has_context={has_context}, has_goto={has_goto}, "
            f"has_find_element={has_find_element}, has_get={has_get}, "
            f"has_current_url={has_current_url}"
        )

    # Check for Playwright first (has locator() method and context)
    if has_locator and has_context and has_goto:
        if is_verbose_enabled():
            print("[ENGINE DETECTION] Detected: playwright")
        return "playwright"

    # Check for Selenium (has find_element and get methods)
    if has_find_element and has_get and has_current_url:
        if is_verbose_enabled():
            print("[ENGINE DETECTION] Detected: selenium")
        return "selenium"

    if is_verbose_enabled():
        print("[ENGINE DETECTION] Could not detect engine!")

    msg = (
        "Unknown browser automation engine. "
        "Expected Playwright Page or Selenium WebDriver object."
    )
    raise ValueError(msg)
